File: src/domain/severity.ts
import { CHECK_ITEMS, type CheckResult, type Checks } from './types';

/**
 * Severity is derived from the checks on every read rather than stored, so a report
 * cannot disagree with itself (DESIGN §1.3).
 */

const RANK: Record<CheckResult, number> = { ok: 0, caution: 1, abnormal: 2 };

export function severityOf(checks: Checks): CheckResult {
  return CHECK_ITEMS.reduce<CheckResult>(
    (worst, item) => (RANK[checks[item]] > RANK[worst] ? checks[item] : worst),
    'ok',
  );
}

export function isAtLeast(result: CheckResult, threshold: CheckResult): boolean {
  return RANK[result] >= RANK[threshold];
}

File: api/app/domain/severity.py
"""Severity is derived from the checks on every read, never stored (DESIGN §1.3)."""

from collections.abc import Mapping

from app.domain.model import CHECK_ITEMS, CHECK_RESULTS, CheckItem, CheckResult


def severity_of(checks: Mapping[CheckItem, CheckResult]) -> CheckResult:
    worst: CheckResult = "ok"
    for item in CHECK_ITEMS:
        if CHECK_RESULTS.index(checks[item]) > CHECK_RESULTS.index(worst):
            worst = checks[item]
    return worst
